extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/kernel/controller/ControllerStageTracker.java

```diff
@@ -280,6 +283,25 @@ ControllerStagePhase addResultKeyStatisticsForWorker(
     return getPhase();
   }
 
+  /**
+   * Sets the {@link #resultPartitions} and {@link #resultPartitionBoundaries} and transitions the phase to POST_READING.
+   */
+  void setClusterByPartitionBoundaries(ClusterByPartitions clusterByPartitions)
+  {
+    if (resultPartitions != null) {
```

Comment:
Lets do a state check that the current state should be MergingStats?

Reply:
And also if this stage.needsKeyStats()

Reply:
Added checks for both